crates/bevy_ecs/src/system/schedule_system.rs

```diff
@@ -0,0 +1,183 @@
+use alloc::borrow::Cow;
+use core::any::TypeId;
+
+use crate::{
+    archetype::ArchetypeComponentId,
+    component::{ComponentId, Tick},
+    query::Access,
+    result::Result,
+    schedule::InternedSystemSet,
+    system::{input::SystemIn, BoxedSystem, System},
+    world::{unsafe_world_cell::UnsafeWorldCell, DeferredWorld, World},
+};
+
+/// A type which wraps and unifies the different sorts of systems that can be added to a schedule.
+pub enum ScheduleSystem {
```

Comment:
I’m not a fan of this approach using an enum as this is introducing a lot of branching. I’d rather see all systems returning a result and figuring out a way of wrapping systems that return ‘()’ to return Ok.

Reply:
I mostly agree. Some considerations:
+ With the enum, I was thinking perhaps the compiler would know which systems are infallible and would eliminate the handler code for them.
+ When writing this, I thought maybe we might eventually want to introduce more variants.

I have since found https://github.com/bevyengine/bevy/pull/8705 and it seems to have a much more streamlined version using adapter systems. I could probably just adopt that. One issue with that version is that it mangles the origin of tracing logs, which is why it was closed. I do not believe the enum does.

Reply:
If you're concerned about branching performance, we should probably just benchmark to see how bad of a regression it really is.

Reply:
We should check against the single threaded executor perf. I wouldn't be surprised if it isn't measurable, but I think we should still go with the other method if we can, since we know we don't need this to be dynamically settable.

Reply:
If it's not a regression, and no one else feels strongly about it, I'd prefer to merge as is and refine the implementation in tree as follow-up. It will be much easier to improve various aspects of this once we have something basic working.

Reply:
Yeah, I don't have strong feelings about the implementation here. Getting the source-logging right is the key element to me.